### ISSUE TITLE
When templating many emails I get a Maximum call stack size exceeded error 

### ISSUE THREAD

**Issue by rmoskal** [mozilla/nunjucks#395]
This is what I get with the latest release:

    { [Template render error: Template render error: (unknown path)
      RangeError: (unknown path)
      RangeError: Maximum call stack size exceeded]
     name: 'Template render error',
     lineno: null,
     colno: null,
     firstUpdate: true,
      Update: [Function] }

I've boiled down the scenario to calling renderString in an async construct like this:

    var nunjucks = require('nunjucks');
    var async = require('async');
    describe('Our email service', function(){
        it.only('regresses to send a LOT of emails', function(done){

            var i=0;

            async.forever(function(cb){
                i++;
                if (i>1000)
                    return cb("done");

                nunjucks.renderString('Hello {{ username }}', { username: 'James' }, cb)

            }, function(err){
                console.log(err, i);
                done()
            });
        });
    });

It happens after 640 successful attempts.  If we use render with an external template then we get this after 565 successful attempts:

    { [Template render error: Template render error: (views/emails/group-milestone.njs)
      RangeError: (views/emails/group-milestone.njs)
      RangeError: Maximum call stack size exceeded]
      name: 'Template render error',
      lineno: null,
      colno: null,
      firstUpdate: true,
      Update: [Function] }

We are able to work around this problem by putting a 1ms delay between templating calls.  Strange, no?

**Comment by devoidfury** [comment]
async.forever
```js
    async.forever = function (fn, callback) {
        function next(err) {
            if (err) {
                if (callback) {
                    return callback(err);
                }
                throw err;
            }
            fn(next);
        }
        next();
    };
```

The fix for this would be to wrap `callback` calls in a setTimeout() to clear the call stack; but whether this should happen in user code, nunjucks, or the async lib is up in the air.

**Comment by rmoskal** [comment]
That's what we did, but it seems strange not to be able to just call the function naively.

**Comment by jlongster** [comment]
The reason is because 95% of users aren't using any of the async stuff, and there's a horrible performance hit when you do so.

We need to make it clear somehow to only use the callback if you are actually using async templates. Otherwise just do `var res = env.renderString(...)`

**Comment by jlongster** [comment]
I realized a way to hide all these ugly internal sync/async details. If you use the async API, your callback is *always* called asynchronously. The key thing is that I was able to detect when the render method is being called internally from nunjucks compiled code, and do careful sync/async work there.

But to the user, now it should be completely hidden. Using the async API, and you're called asynchronously. Always.
